Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift

```diff
@@ -122,6 +122,23 @@ extension NavigationMapView {
         routeRemainingDistancesIndex = allPoints - allRemainingPoints
     }
     
+    func calculateGranularDistanceOffset(_ coordinates: [CLLocationCoordinate2D], splitPoint: CLLocationCoordinate2D) -> Double? {
+        if coordinates.isEmpty { return nil }
+        var distance = 0.0
+        var closestError = Double.greatestFiniteMagnitude
+        var pointDistance: Double? = nil
+        for index in stride(from: coordinates.count - 1, to: 0, by: -1) {
+            let curr = coordinates[index]
+            let prev = coordinates[index - 1]
+            distance += curr.projectedDistance(to: prev)
+            if curr.distance(to: splitPoint) < closestError {
+                pointDistance = distance
+                closestError = curr.distance(to: splitPoint)
+            }
+        }
+        return pointDistance.map { (distance - $0) / distance }
+    }
```

Comment:
This code looks reasonable to me, but we’re gaining a lot of nitty-gritty geometry code here that’s harder to test in isolation. We should move these geometric methods (also the ones below) to a separate file that extends Polyline instead of taking an anonymous array of coordinates. It might even be a good idea to move it to Turf eventually, though we’d need to somehow create a mechanism to plug in alternative projections as needed.

By the way, the map SDK’s Projection class has some utilities specific to Web Mercator that may be useful too.

Reply:
IIRC there was an effort to rework vanishing route line implementation on Maps side, so that we most likely will be able to remove those methods from our SDK completely. So I think there is no urge to improve it right now.

Reply:
Good point, that’s tracked in #4167.